### PR TITLE
fix: avoid duplicate fields in the geometry of the GeoJSON response

### LINKED ISSUE #2248: GeoJSON delivers "type" and "coordinates" twice
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Where did you encounter this issue?

self-hosted instance

### Request URL

https://api.openrouteservice.org/v2/directions/driving-car/geojson

### POST Request Body

```JSON
{"coordinates":[[10.016062, 48.391307], [9.9708714, 48.3508062], [9.9560344, 48.3957671], [9.9993282, 48.4016964], [9.9993282, 48.4016964], [10.0063765, 48.3977211], [10.0431091, 48.4053964]], "preference":"recommended", "language": "de-de", "instructions": "true"}
```

### Response

<details><summary>Response</summary>

```JSON

{"type":"FeatureCollection","bbox":[10.005962,48.39131,10.044105,48.406668],"features":[{"bbox":[10.005962,48.39131,10.044105,48.406668],"type":"Feature","properties":{"segments":[{"distance":3328.5,"duration":480.3,"steps":[{"distance":92.3,"duration":22.1,"type":11,"instruction":"Head northeast on Lupinenweg","name":"Lupinenweg","way_points":[0,3]},{"distance":98.9,"duration":21.3,"type":0,"instruction":"Turn left onto Am Kapellenberg","name":"Am Kapellenberg","way_points":[3,5]},{"distance":1141.8,"duration":115.4,"type":0,"instruction":"Turn left onto Leipheimer Straße, St 2023","name":"Leipheimer Straße, St 2023","way_points":[5,42]},{"distance":48.4,"duration":11.6,"type":7,"instruction":"Enter the roundabout and take the 2nd exit onto Augsburger Straße, St 2023","name":"Augsburger Straße, St 2023","exit_number":2,"way_points":[42,48]},{"distance":1427.1,"duration":185.1,"type":6,"instruction":"Continue straight onto Augsburger Straße, St 2023","name":"Augsburger Straße, St 2023","way_points":[48,81]},{"distance":130.6,"duration":31.3,"type":0,"instruction":"Turn left onto Kantstraße","name":"Kantstraße","way_points":[81,87]},{"distance":362.4,"duration":87.0,"type":1,"instruction":"Turn right onto Offenhauser Straße","name":"Offenhauser Straße","way_points":[87,97]},{"distance":26.9,"duration":6.5,"type":3,"instruction":"Turn sharp right onto Paulstraße","name":"Paulstraße","way_points":[97,98]},{"distance":0.0,"duration":0.0,"type":10,"instruction":"Arrive at Paulstraße, on the right","name":"-","way_points":[98,98]}]},{"distance":1187.9,"duration":211.3,"steps":[{"distance":26.9,"duration":6.5,"type":11,"instruction":"Head southeast on Paulstraße","name":"Paulstraße","way_points":[98,99]},{"distance":74.7,"duration":17.9,"type":1,"instruction":"Turn right onto Kasernstraße","name":"Kasernstraße","way_points":[99,100]},{"distance":828.0,"duration":147.8,"type":0,"instruction":"Turn left onto Reuttier Straße, St 2021","name":"Reuttier Straße, St 2021","way_points":[100,122]},{"distance":177.7,"duration":19.8,"type":0,"instruction":"Turn left onto Finninger Straße","name":"Finninger Straße","way_points":[122,125]},{"distance":43.7,"duration":10.5,"type":1,"instruction":"Turn right onto Nelsonallee","name":"Nelsonallee","way_points":[125,127]},{"distance":37.0,"duration":8.9,"type":0,"instruction":"Turn left","name":"-","way_points":[127,128]},{"distance":0.0,"duration":0.0,"type":10,"instruction":"Arrive at your destination, on the right","name":"-","way_points":[128,128]}]}],"way_points":[0,98,128],"summary":{"distance":4516.4,"duration":691.6}},"geometry":{"coordinates":[[10.043011,48.40552],[10.043079,48.405544],[10.043807,48.405809],[10.044105,48.405919],[10.043592,48.406481],[10.043384,48.406668],[10.042357,48.406456],[10.041859,48.406325],[10.041702,48.406284],[10.040567,48.405995],[10.040264,48.405933],[10.040214,48.405922],[10.039848,48.405869],[10.039407,48.405826],[10.038869,48.405797],[10.038179,48.405734],[10.037898,48.405695],[10.037689,48.40565],[10.037143,48.405525],[10.036979,48.405483],[10.036347,48.405275],[10.036079,48.405185],[10.035721,48.405066],[10.03543,48.404968],[10.035391,48.404955],[10.03473,48.404734],[10.034667,48.404713],[10.03438,48.404631],[10.034084,48.40454],[10.033909,48.404488],[10.033705,48.404425],[10.033148,48.404261],[10.03278,48.404164],[10.032529,48.404093],[10.032504,48.404086],[10.032406,48.404057],[10.031937,48.403916],[10.031156,48.403677],[10.031024,48.403636],[10.030716,48.403538],[10.030596,48.4035],[10.029624,48.403193],[10.029027,48.403007],[10.028987,48.403041],[10.028874,48.403079],[10.028747,48.40307],[10.028646,48.403011],[10.028615,48.402925],[10.02863,48.402881],[10.027541,48.402519],[10.026786,48.402267],[10.026257,48.402117],[10.025565,48.401963],[10.024732,48.401779],[10.024591,48.40175],[10.02443,48.401716],[10.024157,48.401658],[10.023777,48.401581],[10.023503,48.401533],[10.023132,48.401485],[10.022781,48.401457],[10.022223,48.401442],[10.022,48.401455],[10.021803,48.401467],[10.021679,48.401479],[10.021633,48.401483],[10.020831,48.401562],[10.020586,
48.401571],[10.020304,48.401558],[10.01986,48.401507],[10.01936,48.401369],[10.018553,48.401081],[10.018249,48.400973],[10.017878,48.400845],[10.017461,48.400699],[10.016854,48.40055],[10.016535,48.400503],[10.016236,48.400482],[10.01508,48.400444],[10.011566,48.400357],[10.010953,48.40034],[10.010078,48.400318],[10.010091,48.400178],[10.010182,48.400013],[10.010501,48.399603],[10.010558,48.39953],[10.010673,48.399381],[10.010776,48.39925],[10.009903,48.398948],[10.009498,48.398886],[10.00921,48.398871],[10.008951,48.398803],[10.008653,48.398689],[10.007847,48.398342],[10.007612,48.398232],[10.006836,48.397821],[10.00675,48.397749],[10.006722,48.397605],[10.00647,48.39778],[10.006722,48.397605],[10.005962,48.397163],[10.006743,48.396725],[10.006868,48.396646],[10.007504,48.396248],[10.007618,48.396168],[10.008229,48.395723],[10.008925,48.395225],[10.009325,48.394928],[10.009674,48.394652],[10.009706,48.39459],[10.010217,48.394156],[10.010401,48.393999],[10.010855,48.393585],[10.011173,48.393182],[10.011339,48.392951],[10.01158,48.392501],[10.011747,48.392247],[10.011833,48.392168],[10.011905,48.392135],[10.011941,48.392094],[10.01211,48.391895],[10.012366,48.391643],[10.012682,48.39131],[10.013555,48.391567],[10.014241,48.391743],[10.014911,48.39191],[10.01513,48.39157],[10.015145,48.391549],[10.015614,48.391664]],"type":"LineString","coordinates":[[10.043011,48.40552],[10.043079,48.405544],[10.043807,48.405809],[10.044105,48.405919],[10.043592,48.406481],[10.043384,48.406668],[10.042357,48.406456],[10.041859,48.406325],[10.041702,48.406284],[10.040567,48.405995],[10.040264,48.405933],[10.040214,48.405922],[10.039848,48.405869],[10.039407,48.405826],[10.038869,48.405797],[10.038179,48.405734],[10.037898,48.405695],[10.037689,48.40565],[10.037143,48.405525],[10.036979,48.405483],[10.036347,48.405275],[10.036079,48.405185],[10.035721,48.405066],[10.03543,48.404968],[10.035391,48.404955],[10.03473,48.404734],[10.034667,48.404713],[10.03438,48.404631],[10.034084,48.40454],[10.033909,48.404488],[10.033705,48.404425],[10.033148,48.404261],[10.03278,48.404164],[10.032529,48.404093],[10.032504,48.404086],[10.032406,48.404057],[10.031937,48.403916],[10.031156,48.403677],[10.031024,48.403636],[10.030716,48.403538],[10.030596,48.4035],[10.029624,48.403193],[10.029027,48.403007],[10.028987,48.403041],[10.028874,48.403079],[10.028747,48.40307],[10.028646,48.403011],[10.028615,48.402925],[10.02863,48.402881],[10.027541,48.402519],[10.026786,48.402267],[10.026257,48.402117],[10.025565,48.401963],[10.024732,48.401779],[10.024591,48.40175],[10.02443,48.401716],[10.024157,48.401658],[10.023777,48.401581],[10.023503,48.401533],[10.023132,48.401485],[10.022781,48.401457],[10.022223,48.401442],[10.022,48.401455],[10.021803,48.401467],[10.021679,48.401479],[10.021633,48.401483],[10.020831,48.401562],[10.020586,48.401571],[10.020304,48.401558],[10.01986,48.401507],[10.01936,48.401369],[10.018553,48.401081],[10.018249,48.400973],[10.017878,48.400845],[10.017461,48.400699],[10.016854,48.40055],[10.016535,48.400503],[10.016236,48.400482],[10.01508,48.400444],[10.011566,48.400357],[10.010953,48.40034],[10.010078,48.400318],[10.010091,48.400178],[10.010182,48.400013],[10.010501,48.399603],[10.010558,48.39953],[10.010673,48.399381],[10.010776,48.39925],[10.009903,48.398948],[10.009498,48.398886],[10.00921,48.398871],[10.008951,48.398803],[10.008653,48.398689],[10.007847,48.398342],[10.007612,48.398232],[10.006836,48.397821],[10.00675,48.397749],[10.006722,48.397605],[10.00647,48.39778],[10.006722,48.397605],[10.005962,48.397163],[10.006743,48.396725],[10.006868,48.396646],[10.007504,48.396248],[10.007618,48.396168],[10.008229,48.395723],[10.008925,48.395225],[10.009325,48.394928],[10.009674,48.394652],[10.009706,48.39459],[10.010217,48.394156],[10.010401,48.393999],[10.010855,48.393585],[10.011173,48.393182],[10.011339,48.392951],[10.01158,48.392501],[10.011747,48.392247],[10.011833,48.392168],[10.011905,48.392135],[10.011941,48.392094],[10.01211,48.391895],[10.012366,48.391643],[10.012682,48.39131],[10.013555,48.391567],[10.014241,48.391743],[10.
014911,48.39191],[10.01513,48.39157],[10.015145,48.391549],[10.015614,48.391664]],"type":"LineString"}}],"metadata":{"attribution":"openrouteservice.org, OpenStreetMap contributors, tmc - BASt","service":"routing","timestamp":1773769951403,"query":{"coordinates":[[10.0431091,48.4053964],[10.0063765,48.3977211],[10.016062,48.391307]],"profile":"driving-car","profileName":"driving-car","format":"geojson"},"engine":{"version":"9.7.1","build_date":"2026-02-27T11:54:37Z","graph_date":"2026-03-13T15:42:02Z","osm_date":"2026-01-06T21:21:12Z"}}}

```
</details>

### Current behavior

Since Version 9.6.0 OpenRouteService delivers the fields „type“ and „coordinates“ twice with identical content.

### Expected behavior

The fields „type“ and „coordinates“ appear only once.

### Openrouteservice Version

9.7.1

### Build date

2026-02-27T11:54:37Z

### Graph date

2026-03-13T15:42:02Z

### Forum Topic Link

_No response_

### PR DESCRIPTION
**Pull request by aoles**
### Pull Request Checklist

- [x] 1. I have [**rebased**][rebase] the latest version of the main branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [ ] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [x] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [x] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [x] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [ ] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes #2248.

### Information about the changes
- Key functionality added: address duplicate fields.
- Reason for change: invalid GeoJSON response.

[config]: https://GIScience.github.io/openrouteservice/run-instance/configuration/
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/main/CONTRIBUTE.md#pull-request-guidelines

### CODE REVIEW

File: ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java [hunk 1]
```diff
@@ -14,35 +14,20 @@
...
     @JsonProperty("coordinates")
-    public Object getCoordinates() {
-        return json.get("coordinates");
-    }
+    public final Object coordinates;
 
-    @JsonAnyGetter
-    public JSONObject getJson() {
-        return json;
+    public GeoJSONGeometry(String type, Object coordinates) {
```
Comment (mmerdes) [inline ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java]: Could `coordinates` be typed more speficially?
Maybe as `org.json.simple.JSONArray` or at least `org.json.simple.JSONAware`?

File: ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java [hunk 2]
```diff
@@ -14,35 +14,20 @@
 
 package org.heigit.ors.api.responses.routing.geojson;
 
-import com.fasterxml.jackson.annotation.JsonAnyGetter;
 import com.fasterxml.jackson.annotation.JsonProperty;
```
Comment (mmerdes) [inline ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java:17]: Could we use the newer Jackson versions here (package `tools.jackson`)?

File: ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java [hunk 3]
```diff
@@ -14,35 +14,20 @@
...
 import io.swagger.v3.oas.annotations.media.Schema;
-import org.json.simple.JSONObject;
 
 public class GeoJSONGeometry {
```
Comment (mmerdes) [inline ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java]: Would the Jackson-annotations still work with a Java record?
If so, using a record might be a bit nicer ;)

File: ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java [hunk 4]
```diff
@@ -752,6 +756,33 @@ void testGeoJsonExport() {
...
+                .post(getEndPointPath() + "/{profile}/" + format)
+                .then()
+                .assertThat()
+                .statusCode(200)
+                .extract()
+                .response();
+
+        // Configure Jackson for strict duplicate detection
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.enable(JsonParser.Feature.STRICT_DUPLICATE_DETECTION);
```
Comment (mmerdes) [inline ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java]: If something is wrong with the 'strict duplicate detection' we would have a silent error masking.

How about creating the `strictObjectMapper` in a factory method and adding a small test that `shouldFail()` given a string with duplicates?

File: ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java [hunk 5]
```diff
@@ -752,6 +756,33 @@ void testGeoJsonExport() {
...
+
+        // Attempt to parse the JSON and fail if any exception is thrown
+        Assertions.assertDoesNotThrow(() -> mapper.readTree(response.asString()));
```
Comment (mmerdes) [inline ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java]: Statically importing `assertDoesNotThrow` would make it a tiny bit more idiomatic ;)

File: ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java (hunk 4 above)
Comment (mmerdes) [reply ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java]: Alternatively, we could just add the same small check **within** the existing test after creating the mapper.

**Review by mmerdes — COMMENTED** [review]
Added a few minor suggestions.

File: ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java (hunk 2 above)
Comment (aoles) [reply ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java:17]: In principle yes. However, as we already rely on `com.fasterxml.jackson.*` in 70 other classes (!), upgrading it across the entire codebase in a separate PR would be the preferred approach.

File: ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java (hunk 5 above)
Comment (aoles) [reply ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java]: Agreed! 👍

**Comment by aoles** [comment]
Thanks a lot @mmerdes for your thorough review! 🧐 

I’ve updated the PR according to your suggestions. I hope I was able to address most of your concerns.

Looking forward to your feedback 🤓 

Cheers!

File: ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java (hunk 4 above)
Comment (aoles) [reply ors-api/src/test/java/org/heigit/ors/apitests/routing/ResultTest.java]: Fair point - done! 🚀

File: ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java (hunk 3 above)
Comment (aoles) [reply ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java]: Nice idea, seems to work! 🎉

File: ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java (hunk 1 above)
Comment (aoles) [reply ors-api/src/main/java/org/heigit/ors/api/responses/routing/geojson/GeoJSONGeometry.java]: Fair enough, fixed 👍

**Review by mmerdes — APPROVED** [review]
Thanks for working through the comments very thoroughly @aoles.

LGTM now!
